### PR TITLE
fix(NODE-4429): select server sync for endSessions during close

### PR DESCRIPTION
**Pull request by durran**
### Description

This prevents the `endSessions` command from going through server selection when no server would be selected after calling `client.close()`. This could happen in situations where the server has gone down but server selection keeps forcing monitor checks for up to `serverSelectionTimeoutMS` when an immediate return is desired.

#### What is changing?

- `MongoClient`'s `close()` method now does a quick server selection to get the list of servers that would be returned with a primary preferred read preference. If nothing is returned then calling `endSessions` is skipped.

##### Is there new documentation needed for these changes?

None

#### What is the motivation for this change?

NODE-4429

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

File: test/integration/node-specific/mongo_client.test.ts [hunk 1]
```diff
@@ -649,5 +650,25 @@ describe('class MongoClient', function () {
       expect(endEvents).to.have.lengthOf(1);
       expect(endEvents[0]).to.have.property('reply', undefined); // noReponse: true
     });
+
+    context('when server selection would return no servers', () => {
+      const serverDescription = new ServerDescription('a:1');
+
+      it('short circuits and does not end sessions', async () => {
```
Comment (durran) [inline test/integration/node-specific/mongo_client.test.ts:657]: This test will hang and wait for server selection without the fix in, thus reproducing the path.

Comment (nbbeeken) [reply test/integration/node-specific/mongo_client.test.ts:657]: Just to demonstrate that `endSession` does not do anything but bookkeeping, let's assert that the `client.s.sessionPool` still have one session in it both before and after close.

File: src/mongo_client.ts [hunk 2]
```diff
@@ -500,6 +501,16 @@ export class MongoClient extends TypedEventEmitter<MongoClientEvents> {
...
         .then(() => {
+          // If we would attempt to select a server and get nothing back we short circuit
+          // to avoid the server selection timeout.
+          const selector = readPreferenceServerSelector(ReadPreference.primaryPreferred);
+          const topologyDescription = this.topology.description;
+          const serverDescriptions = Array.from(topologyDescription.servers.values());
+          const servers = selector(topologyDescription, serverDescriptions);
+          if (servers.length === 0) {
+            return callback();
+          }
```
Comment (nbbeeken) [inline src/mongo_client.ts:512]: I think we want to return here, could be wrong, but the callback should only be referenced in the last "then" of the chain. 

So actually a few lines down where we do: 
```ts
if (this.topology == null) {
  return callback();
}
```
Is a teeny bug

**Review by nbbeeken — CHANGES_REQUESTED** [review]

File: src/mongo_client.ts (hunk 2 above)
Comment (durran) [reply src/mongo_client.ts]: I've updated to assert there is only 1 session in the pool and changed the check to just return.

File: test/integration/node-specific/mongo_client.test.ts (hunk 1 above)
Comment (durran) [reply test/integration/node-specific/mongo_client.test.ts:657]: Added.

**Review by nbbeeken — APPROVED** [review]
Small TS error I think but LGTM otherwise

File: src/mongo_client.ts [hunk 3]
```diff
@@ -500,6 +501,16 @@ export class MongoClient extends TypedEventEmitter<MongoClientEvents> {
 
       Promise.all(activeSessionEnds)
         .then(() => {
+          // If we would attempt to select a server and get nothing back we short circuit
+          // to avoid the server selection timeout.
+          const selector = readPreferenceServerSelector(ReadPreference.primaryPreferred);
+          const topologyDescription = this.topology.description;
```
Comment (dariakp) [inline src/mongo_client.ts:507]: Can `this.topology` be undefined here? If so, what would be the correct way to handle that?

File: test/integration/node-specific/mongo_client.test.ts [hunk 4]
```diff
@@ -649,5 +650,26 @@ describe('class MongoClient', function () {
       expect(endEvents).to.have.lengthOf(1);
       expect(endEvents[0]).to.have.property('reply', undefined); // noReponse: true
     });
+
+    context('when server selection would return no servers', () => {
```
Comment (dariakp) [inline test/integration/node-specific/mongo_client.test.ts:654]: Given the other comment about possibly undefined topology, it would be good to also cover that edge case, too, if we can

**Review by dariakp — CHANGES_REQUESTED** [review]

File: src/mongo_client.ts (hunk 3 above)
Comment (durran) [reply src/mongo_client.ts:510]: I've add the check after to address this.

File: test/integration/node-specific/mongo_client.test.ts (hunk 4 above)
Comment (durran) [reply test/integration/node-specific/mongo_client.test.ts:654]: The test case to cover this would be quite involved to get the driver in an internal state where the topology was actually null at that point in the code. The only situation where it could happen would be simultaneous async calls to `close()` where one had gotten to the point of closing the topology before the other one got to the point of ending sessions. I started to try and write a test and it just got way out of hand trying to mimic that behaviour in it.

Comment (dariakp) [reply test/integration/node-specific/mongo_client.test.ts:654]: ok, given it's an edge case I think it's not a big deal, and we're just following the pattern in the .then chain

leaving the comment open for visibility

**Review by dariakp — APPROVED** [review]

**Review by baileympearson — APPROVED** [review]

**Review by nbbeeken — APPROVED** [review]
